datafusion/core/src/datasource/file_format/parquet.rs

```diff
@@ -719,55 +717,300 @@ impl DataSink for ParquetSink {
+) -> Result<usize> {
+    let mut row_count = 0;
+    let parallelism = data.len();
+    let mut join_handles: Vec<JoinHandle<ParquetFileSerializedResult>> =
+        Vec::with_capacity(parallelism);
+    for _ in 0..parallelism {
+        let buffer: Vec<u8> = Vec::new();
+        let mut writer = parquet::arrow::arrow_writer::ArrowWriter::try_new(
+            buffer,
+            output_schema.clone(),
+            Some(parquet_props.clone()),
+        )?;
+        let mut data_stream = data.remove(0);
+        join_handles.push(tokio::spawn(async move {
+            let mut inner_row_count = 0;
+            while let Some(batch) = data_stream.next().await.transpose()? {
+                inner_row_count += batch.num_rows();
+                writer.write(&batch)?;
+            }
+            let out = writer.into_inner()?;
+            Ok((out, inner_row_count))
+        }))
+    }
+
+    let mut writer = None;
+    let shared_buff = SharedBuffer::new(1048576);
+    for handle in join_handles {
+        let join_result = handle.await;
+        match join_result {
+            Ok(result) => {
+                let (out, num_rows) = result?;
+                let reader = bytes::Bytes::from(out);
+                row_count += num_rows;
+                //let reader = File::open(buffer)?;
+                let metadata = parquet::file::footer::parse_metadata(&reader)?;
+                let schema = metadata.file_metadata().schema();
+                writer = match writer {
+                    Some(writer) => Some(writer),
+                    None => Some(SerializedFileWriter::new(
+                        shared_buff.clone(),
+                        Arc::new(schema.clone()),
+                        Arc::new(parquet_props.clone()),
+                    )?),
+                };
+
+                match &mut writer{
+                    Some(w) => {
+                        // Note: cannot use .await within this loop as RowGroupMetaData is not Send
+                        // A tokio::sync::mpsc::unbounded_channel works, but it is expensive and slow
+                        // to drain(..) the bytes and send through the buffer on each loop. Plus
+                        // it results in only a small improvement in max memory usage vs. flushing
+                        // after RowGroupMetaData goes out of scope.
+                        for rg in metadata.row_groups() {
+                            let mut rg_out = w.next_row_group()?;
+                            for column in rg.columns() {
+                                let result = ColumnCloseResult {
+                                    bytes_written: column.compressed_size() as _,
+                                    rows_written: rg.num_rows() as _,
+                                    metadata: column.clone(),
+                                    bloom_filter: None,
+                                    column_index: None,
+                                    offset_index: None,
```

Comment:
These are really important for query performance, so we probably want to find a way to support these. This may require modifying some of the upstream APIs if they don't expose the necessary information

Reply:
Filed https://github.com/apache/arrow-datafusion/issues/7589 and https://github.com/apache/arrow-datafusion/issues/7590 to track -- I'll also add a code level comment